x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Case.java

```diff
@@ -0,0 +1,155 @@
+import static org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+/**
+ * Implements the CASE WHEN ... THEN ... ELSE ... END expression
+ */
+public class Case extends ConditionalFunction {
+
+    private final List<IfConditional> conditions;
+    private final Expression defaultElse;
+
+    @SuppressWarnings("unchecked")
+    public Case(Source source, List<Expression> expressions) {
+        super(source, expressions);
+        this.conditions = (List<IfConditional>) (List<?>) expressions.subList(0, expressions.size() - 1);
+        this.defaultElse = expressions.get(expressions.size() - 1);
+        setDataType();
+    }
+
+    public List<IfConditional> conditions() {
+        return conditions;
+    }
+
+    public Expression defaultElse() {
+        return defaultElse;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new Case(source(), newChildren);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, Case::new, children());
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        DataType resultDataType;
+        if (conditions.isEmpty()) {
+            resultDataType = defaultElse().dataType();
+        } else {
+            resultDataType = conditions.get(0).dataType();
+        }
+
+        for (IfConditional conditional : conditions) {
+            if (conditional.condition().dataType() != DataType.BOOLEAN) {
+                return new TypeResolution(format(null, "condition of [{}] must be [boolean], found value [{}] type [{}]",
+                    conditional.sourceText(),
+                    Expressions.name(conditional.condition()),
+                    conditional.condition().dataType().typeName));
+            }
+            if (DataTypes.areTypesCompatible(resultDataType, conditional.dataType()) == false) {
+                return new TypeResolution(format(null, "result of [{}] must be [{}], found value [{}] type [{}]",
+                    conditional.sourceText(),
+                    resultDataType.typeName,
+                    Expressions.name(conditional.result()),
+                    conditional.dataType().typeName));
+            }
+        }
+
+        if (DataTypes.areTypesCompatible(dataType, defaultElse.dataType()) == false) {
+            return new TypeResolution(format(null, "ELSE clause of [{}] must be [{}], found value [{}] type [{}]",
+                defaultElse.sourceText(),
+                resultDataType.typeName,
+                Expressions.name(defaultElse),
+                defaultElse.dataType().typeName));
+        }
+
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public boolean foldable() {
+        return defaultElse.foldable() && (conditions.isEmpty() ||
+            (conditions.size() == 1 && conditions.get(0).condition().foldable() && conditions.get(0).result().foldable()));
```

Comment:
I'm not sure this method is correct - why is the defaultElse checked first - if a condition is foldable and true, `defaultElse` is not relevant.
I think it should be:
```java
return (conditions.isEmpty() && defaultElse.foldable()) || (conditions.size() > 0 && conditions.get(0).condition().foldable());
```

Reply:
Actually it should be: 
```
return (conditions.isEmpty() && defaultElse.foldable()) ||
            (conditions.isEmpty() == false && conditions.get(0).condition().foldable() && conditions.get(0).result().foldable());
```

since we know that the optimizer has removed all the FALSE conditions and the result should also be foldable.

Reply:
we cannot say: `conditions.size() > 0` cause this will cause `ConstantFolding` rule to get applied before the `SimplifyCase` takes place.